From 02626a83a89068944e8840fef9e6deada9ae67b1 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Sat, 31 Jan 2026 13:41:36 +0000
Subject: [PATCH] server/sockets.c: Print the actual bound addresses and ports
 in debug output

Previously when listening on a TCP/IP port, we printed (in debug
output) this:

  nbdkit: debug: bound to IP address <any>:10809 (2 socket(s))

Change this to print the actual socket addresses for each socket:

  nbdkit: debug: bound to IPv4 address 0.0.0.0:10809
  nbdkit: debug: bound to IPv6 address [::]:10809

We just print the raw IP addresses to avoid doing name lookups.  This
is still way harder than it should be.

(cherry picked from commit 6ef03ffdd451b730ac90d889e758e9151b14db82)
---
 configure.ac     |  7 ++++++
 server/sockets.c | 57 ++++++++++++++++++++++++++++++++++++++++++++++--
 2 files changed, 62 insertions(+), 2 deletions(-)

diff --git a/configure.ac b/configure.ac
index a7a021e0..1cc87a44 100644
--- a/configure.ac
+++ b/configure.ac
@@ -509,6 +509,13 @@ AC_CHECK_MEMBERS([struct sockpeercred.uid], [], [],
 #endif
 ]])
 
+dnl Check for types.
+AC_CHECK_TYPES([in_port_t], [], [], [[
+#ifdef HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+]])
+
 dnl Replacement functions that we provide for some platforms.
 AC_CONFIG_LIBOBJ_DIR([common/replacements])
 AC_REPLACE_FUNCS([\
diff --git a/server/sockets.c b/server/sockets.c
index 348da029..cef27d57 100644
--- a/server/sockets.c
+++ b/server/sockets.c
@@ -46,6 +46,10 @@
 #include <sys/socket.h>
 #endif
 
+#ifdef HAVE_ARPA_INET_H
+#include <arpa/inet.h>
+#endif
+
 #ifdef HAVE_SYS_UN_H
 #include <sys/un.h>
 #endif
@@ -271,8 +275,57 @@ bind_tcpip_socket (sockets *socks)
     exit (EXIT_FAILURE);
   }
 
-  debug ("bound to IP address %s:%s (%zu socket(s))",
-         ipaddr ? ipaddr : "<any>", ipport, socks->len);
+  /* Print the IP address of each socket in debug output. */
+  if (verbose) {
+    char addr_str[128];
+    size_t i;
+    struct sockaddr_storage ss;
+    socklen_t sslen = sizeof ss;
+    const struct sockaddr_in *sin;
+    const struct sockaddr_in6 *sin6;
+    const void *addr;
+    const char *familyname;
+#ifdef HAVE_IN_PORT_T
+    in_port_t portno;
+#else
+    int portno;
+#endif
+    bool square_brackets = false;
+
+    for (i = 0; i < socks->len; ++i) {
+      if (getsockname (socks->ptr[i], (struct sockaddr *) &ss, &sslen) == -1) {
+        debug ("getsockname: %m");
+        continue;
+      }
+      switch (ss.ss_family) {
+      case AF_INET:
+        familyname = "IPv4";
+        sin = (const struct sockaddr_in *) &ss;
+        addr = &sin->sin_addr;
+        portno = sin->sin_port;
+        break;
+      case AF_INET6:
+        familyname = "IPv6";
+        square_brackets = true;
+        sin6 = (const struct sockaddr_in6 *) &ss;
+        addr = &sin6->sin6_addr;
+        portno = sin6->sin6_port;
+        break;
+      default:
+        /* Probably can never happen? */
+        debug ("unknown address family: %d", ss.ss_family);
+        continue;
+      }
+      if (inet_ntop (ss.ss_family, addr, addr_str, sizeof addr_str) == NULL) {
+        debug ("inet_ntop: %m");
+        continue;
+      }
+      debug ("bound to %s address %s%s%s:%d",
+             familyname,
+             square_brackets ? "[" : "", addr_str, square_brackets ? "]" : "",
+             (int) ntohs (portno));
+    }
+  }
 }
 
 void
-- 
2.47.3

